### PR TITLE
#145 For symmetry, define a logicalLocation object

### PR DESCRIPTION
**Pull request by EasyRhin0**
Hope I understood the fullyQualifiedName correctly...

### CODE REVIEW

File: src/Sarif.Converters/PREFastConverter.cs [hunk 1]
```diff
@@ -99,6 +106,17 @@ private Result CreateResult(Defect defect)
             return result;
         }
 
+        private void TryAddLogicalLocation(string name, string decoratedName)
```
Comment (lgolding) [inline src/Sarif.Converters/PREFastConverter.cs]: This isn't a `Try` function -- those return `false` if they can't do what you ask. This is just an overload of `AddLogicalLocation`.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif [hunk 2]
```diff
@@ -12,6 +12,12 @@
           "mimeType": "text/x-cpp"
         }
       },
+      "logicalLocations": {
+        "C6385_IndexAlias": {
+          "name": "C6385_IndexAlias",
+          "decoratedName": "?C6385_IndexAlias@@YAXH@Z"
+        }
+      },
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif]: Yeah, that looks right. Strictly speaking you don't need `name` if it's the same as the key. I'll keep reading and decide if this is worth fixing.

File: src/Sarif.Converters/ToolFileConverterBase.cs [hunk 3]
```diff
@@ -49,6 +49,11 @@ protected internal string AddLogicalLocation(LogicalLocation logicalLocation, st
                 ++disambiguator;
             }
 
+            if (disambiguator > 0)
+            {
+                logicalLocation.FullyQualifiedName = logicalLocationKey;
```
Comment (lgolding) [inline src/Sarif.Converters/ToolFileConverterBase.cs]: Bug. You correctly determined that the FQLN (`"A.B.C"`) is different from the key (`"A.B.C-1"` or whatever). But then you need to set `logicalLocation.FullyQualifiedName` to the FQLN, _not_ to the key.

As the code stands, you'll get:
```
"logicalLocations": {
  "A.B.C-1": {
    "fullyQualifiedName": "A.B.C-1"
   }
}
```
You want:
```
"logicalLocations": {
  "A.B.C-1": {
    "fullyQualifiedName": "A.B.C"
   }
}
```
Since your tests are passing, it's clear we never had a test for this corner case. Now's the time to add one. It would show that if you add two different logical locations with the same FQLN to the `logicalLocations` dictionary, you'll get two dictionary elements with different keys but the same `fullyQualifiedName`.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif [hunk 4]
```diff
@@ -12,6 +12,53 @@
...
+          "name": "foo",
+          "fullyQualifiedName": "foo",
+          "decoratedName": "?foo@@YADXZ"
+        },
+        "bar": {
+          "name": "bar",
+          "decoratedName": "?bar@@YAXPAHII@Z"
+        },
+        "ShiftTest::f": {
+          "name": "ShiftTest::f",
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: Bug: `name` should be only the last component, `f`. That's what distinguishes it from the "fully qualified name" (`ShiftTest::f`).

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif (hunk 2 above)
Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif]: Yes, do fix this. The spec says:
>If the `logicalLocation` object describes a top-level logical location, and if the `name` property would equal the name of the property for which this object provides the value, then the name property **MAY** be absent.
So what you have here is legal SARIF, but our code should produce "best" SARIF, and I think it best not to emit unnecessary properties.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-101.xml.sarif [hunk 5]
```diff
@@ -12,6 +12,24 @@
...
+        "bad1": {
+          "name": "bad1",
+          "decoratedName": "?bad1@@YAXPAHIPAU_FLEXARRAY@@@Z"
+        },
+        "bad2": {
+          "name": "bad2",
+          "decoratedName": "?bad2@@YAPAU_FLEXARRAY@@I@Z"
+        },
+        "Foo::what": {
+          "name": "Foo::what",
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-101.xml.sarif]: Just to emphasize: Here `name` is required, because it should be `"what"`, which doesn't equal the key `"Foo::what"`.

File: src/Sarif.Viewer.VisualStudio/Sarif/Run.extensions.cs [hunk 6]
```diff
@@ -34,21 +34,14 @@ static JObject RuleMetadata
             }
         }
 
-        public static bool TryGetRule(this Run run, string ruleId, string ruleKey, out IRule rule)
+        public static bool TryGetRule(this Run run, string ruleId, out IRule rule)
```
Comment (lgolding) [inline src/Sarif.Viewer.VisualStudio/Sarif/Run.extensions.cs]: Yes, much simpler now. That was the idea of this change.

File: src/Sarif/Baseline/DefaultBaseline/LocationBaselineEquals.cs [hunk 7]
```diff
@@ -19,8 +19,8 @@ public bool Equals(Location x, Location y)
                     return false;
                 }
 
-                // Code locations (decorated name/fully qualified logical name) should match.
-                if (x.DecoratedName != y.DecoratedName || x.FullyQualifiedLogicalName != y.FullyQualifiedLogicalName)
+                // Code locations (fully qualified logical name) should match.
```
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Sarif/Baseline/DefaultBaseline/LocationBaselineEquals.cs]: I mean, "file an issue" :)

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif (hunk 4 above)
Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: BTW, this might be a bug in the PREfast converter. If so, file it and move on.

File: src/Sarif.Converters/PREFastConverter.cs (hunk 1 above)
Comment (EasyRhin0) [reply src/Sarif.Converters/PREFastConverter.cs]: Fixed

File: src/Sarif.Converters/ToolFileConverterBase.cs (hunk 3 above)
Comment (EasyRhin0) [reply src/Sarif.Converters/ToolFileConverterBase.cs]: But the disambiguated string is in generatedKey. logicalLocationKey is the original unchanged value

File: src/Sarif/Baseline/DefaultBaseline/LocationBaselineEquals.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif/Baseline/DefaultBaseline/LocationBaselineEquals.cs]: OK

File: src/Sarif.Converters/ToolFileConverterBase.cs (hunk 3 above)
Comment (lgolding) [reply src/Sarif.Converters/ToolFileConverterBase.cs]: Oh! Then consider whether we should rename it.

Comment (EasyRhin0) [reply src/Sarif.Converters/ToolFileConverterBase.cs]: string fullyQualifiedLogicalName ... ?

Comment (lgolding) [reply src/Sarif.Converters/ToolFileConverterBase.cs]: Sounds good!

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif (hunk 4 above)
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: Yeah I haven't touched this

Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: Thanks.

@michaelcfanning  FYI there's a bug in the PREfast converter that populates `logicalLocation.name` with the FQLN, as in this example. @EasyRhin0 will file it.

File: src/Sarif/Baseline/DefaultBaseline/LocationBaselineEquals.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif/Baseline/DefaultBaseline/LocationBaselineEquals.cs]: Done #887

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif (hunk 2 above)
Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif]: @michaelcfanning This is another non-optimal behavior in the PREfast converter (not strictly a bug) where it populates `logicalLocation.name` when it doesn't need to. @EasyRhin0 please file.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif (hunk 4 above)
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: Actually I did add this to persist the decoratedName, so I'll fix.

Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: In that case... how sure are we that the other issue, where we populate `name` unnecessarily, is a converter issue? Could that also be your bug?

Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: Yeah same, though other converters have the same problem. FxCop:
        "wpfapplication5.exe": {
          "name": "wpfapplication5.exe",
          "fullyQualifiedName": "wpfapplication5.exe",
          "kind": "module"
        },

In the PREfast case, the Defect.Function property is used and it is the full string. Should I be parsing based on "::"?

Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: Ah never mind I think I have it.

Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: (Yes, split on `::`.)

Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: @EasyRhin0 @michaelcfanning File a separate bug for any converter with this non-optimal behavior.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif (hunk 2 above)
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/AA_IndexAlias.cpp.xml.sarif]: Fixed

**Comment by EasyRhin0** [comment]
Need to fix some unit tests...

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-016-amd64.xml.sarif [hunk 8]
```diff
@@ -12,6 +12,19 @@
           "mimeType": "text/x-cpp"
         }
       },
+      "logicalLocations": {
+        "C::Bad1": {
+          "name": "Bad1",
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-016-amd64.xml.sarif]: Nice!!!

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif [hunk 9]
```diff
@@ -12,6 +12,46 @@
...
+          "decoratedName": "?BugLast@@YADPBD@Z"
+        },
+        "g1": {
+          "decoratedName": "?g1@@YAXPAH@Z"
+        },
+        "foo-0": {
+          "name": "foo",
+          "fullyQualifiedName": "foo",
+          "decoratedName": "?foo@@YADXZ"
+        },
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif]: Perfect. There really are two different `foo`s (different decorated names), and this one has the extra properties it needs.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-050.xml.sarif [hunk 10]
```diff
@@ -12,6 +12,16 @@
...
+      "logicalLocations": {
+        "f": {
+          "decoratedName": "?f@@YAXH@Z"
+        },
+        "f-0": {
+          "name": "f",
+          "fullyQualifiedName": "f",
+          "decoratedName": "?f@@YAXPAUS@@@Z"
+        }
+      },
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-050.xml.sarif]: So nice.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-049.xml.sarif [hunk 11]
```diff
@@ -12,6 +12,20 @@
           "mimeType": "text/x-cpp"
         }
       },
+      "logicalLocations": {
+        "": {
+          "decoratedName": ""
+        },
```
[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-049.xml.sarif]: Ah didn't notice that. Will investigate.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-071.xml.sarif [hunk 12]
```diff
@@ -12,6 +12,34 @@
...
+          "decoratedName": "?GetComputerName@@YAXPADPAI@Z"
+        },
+        "TestFields": {
+          "decoratedName": "?TestFields@@YAXPADHH@Z"
+        },
+        "f-0": {
+          "name": "f",
+          "fullyQualifiedName": "f",
+          "decoratedName": "?f@@YAXPADPAI@Z"
+        },
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-071.xml.sarif]: I love this every time I see it :-)

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-049.xml.sarif (hunk 11 above)
[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-049.xml.sarif]: There is an item that is mostly empty:
<FUNCTION></FUNCTION><DECORATED></DECORATED><FUNCLINE>0</FUNCLINE>

I don't know if this is a valid PREfast defect but I guess we should omit it on conversion.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-071.xml.sarif (hunk 12 above)
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-071.xml.sarif]: Awesome :)

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-049.xml.sarif (hunk 11 above)
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-049.xml.sarif]: Got it. I think the converter does the right thing here. This may be bad PREfast, but that's not our business. Let's leave the code as is.

[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-049.xml.sarif]: Roger.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-101.xml.sarif [hunk 13]
```diff
@@ -12,6 +12,21 @@
...
       },
+      "logicalLocations": {
+        "bad1": {
+          "decoratedName": "?bad1@@YAXPAHIPAU_FLEXARRAY@@@Z"
+        },
+        "bad2": {
+          "decoratedName": "?bad2@@YAPAU_FLEXARRAY@@I@Z"
+        },
+        "Foo::what": {
+          "name": "what",
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-101.xml.sarif]: Yes, this is the one that originally caught my eye. Funny how I missed it so many times in the preceding files.

File: src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-103.xml.sarif [hunk 14]
```diff
@@ -12,6 +12,34 @@
...
+          "decoratedName": "?f1@@YAXXZ"
+        },
+        "g": {
+          "decoratedName": "?g@@YAXPAUUS@@@Z"
+        },
+        "g-0": {
+          "name": "g",
+          "fullyQualifiedName": "g",
+          "decoratedName": "?g@@YAXXZ"
+        },
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-103.xml.sarif]: And again. <3

File: src/Sarif.FunctionalTests/v2/ConverterTestData/StaticDriverVerifier/cancelspinlock_bug1.tt.sarif [hunk 15]
```diff
@@ -346,7 +346,7 @@
                       "nestingLevel": 2,
                       "importance": "unimportant",
                       "properties": {
-                        "currentDataValues": "^&sdv_ControllerIrp=8520",
+                        "currentDataValues": "^\u0026sdv_ControllerIrp=8520",
```
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/StaticDriverVerifier/cancelspinlock_bug1.tt.sarif]: Don't get hung up on it, just curious.

**Comment by EasyRhin0** [comment]
What was the suggestion? Not seeing it

File: src/Sarif.FunctionalTests/v2/ConverterTestData/StaticDriverVerifier/cancelspinlock_bug1.tt.sarif (hunk 15 above)
[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/StaticDriverVerifier/cancelspinlock_bug1.tt.sarif]: I'm going to file a bug on this voodoo.

[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/v2/ConverterTestData/StaticDriverVerifier/cancelspinlock_bug1.tt.sarif]: #895
